src/shogun/mathematics/linalg/LinalgNamespace.h

```diff
@@ -468,6 +495,63 @@ T dot(const SGVector<T>& a, const SGVector<T>& b)
 	return infer_backend(a, b)->dot(a, b);
 }
 
+/**
+ * Compute the eigenvalues and eigenvectors of a matrix.
+ * Note that the type of the computed values is the same
+ * as the matrix's type, i.e. for real matrices it returns
+ * only the real part of the eigenvalues/vectors.
+ *
+ * User should pass an appropriately pre-allocated memory vector
+ * to store the eigenvalues and an appropriately pre-allocated memory
+ * matrix to store the eigenvectors.
+ *
+ * @param A The matrix whose eigenvalues and eigenvectors are to be computed
+ * @param eigenvalues Eigenvalues result vector
+ * @param eigenvectors Eigenvectors result matrix
+ */
+template <typename T>
+void eigen_solver(const SGMatrix<T>& A, SGVector<T>& eigenvalues, SGMatrix<T>& eigenvectors)
+{
+	REQUIRE(A.num_rows == A.num_cols, "Matrix A is not square!\n");
+	REQUIRE(A.num_rows == eigenvectors.num_rows,
+			"Number of rows of A (%d) doesn't match eigenvectors' matrix (%d).\n",
+		A.num_rows, eigenvectors.num_rows);
+	REQUIRE(A.num_cols == eigenvectors.num_cols,
+			"Number of columns of A (%d) doesn't match eigenvectors' matrix (%d).\n",
+			A.num_cols, eigenvectors.num_cols);
+	REQUIRE(A.num_cols == eigenvalues.vlen,
+			"Length of eigenvalues' vector doesn't match matrix A");
+
+	infer_backend(A)->eigen_solver(A, eigenvalues, eigenvectors);
+}
+
+/**
+ * Compute the eigenvalues and eigenvectors of a symmetric matrix.
+ *
+ * User should pass an appropriately pre-allocated memory vector
+ * to store the eigenvalues and an appropriately pre-allocated memory
+ * matrix to store the eigenvectors.
+ *
+ * @param A The matrix whose eigenvalues and eigenvectors are to be computed
+ * @param eigenvalues Eigenvalues result vector
+ * @param eigenvectors Eigenvectors result matrix
+ */
+template <typename T>
+void eigen_solver_symmetric(const SGMatrix<T>& A, SGVector<T>& eigenvalues, SGMatrix<T>& eigenvectors)
+{
+	REQUIRE(A.num_rows == A.num_cols, "Matrix A is not square!\n");
+	REQUIRE(A.num_rows == eigenvectors.num_rows,
+			"Number of rows of A (%d) doesn't match eigenvectors' matrix (%d).\n",
+			A.num_rows, eigenvectors.num_rows);
+	REQUIRE(A.num_cols == eigenvectors.num_cols,
```

Comment:
@vigsterkr (and everyone) I feel like we should put all those checks into functions so that we can unify their formatting.
Any thoughts on that @micmn ?